### PR TITLE
Add more supported types to SliceBase

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds more supported types to SliceBase operator, needed to be able to slice shapes which are typical int64

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Added more types to SliceBase*
     *Added a clamp specialization for uint64, since half doesn't have the necessary conversion constructor*
 - Affected modules and functionalities:
     *Slice and Crop*
 - Key points relevant for the review:
     *N/A*
 - Validation and testing:
     *N/A*
 - Documentation (including examples):
     *N/A*

**JIRA TASK**: *[Use DALI-XXXX or NA]*

### CODE REVIEW

File: dali/operators/generic/slice/slice_base.h [hunk 1]
```diff
@@ -32,7 +32,9 @@
 #include "dali/pipeline/util/operator_impl_utils.h"
 #include "dali/util/crop_window.h"
 
-#define SLICE_TYPES (uint8_t, int16_t, uint16_t, int32_t, float, float16)
+#define SLICE_TYPES (uint8_t, uint16_t, uint32_t, uint64_t, \
+                     int8_t,  int16_t,  int32_t,  int64_t, \
+                     float, float16)
```
Comment (klecki) [inline dali/operators/generic/slice/slice_base.h:37]:
```suggestion
                     float16, float, double)
```

File: dali/pipeline/data/types.h [hunk 2]
```diff
@@ -374,21 +461,6 @@ class DLL_PUBLIC TypeTable {
...
-    return "list of " + TypeTable::GetTypeName<T>();
-  }
-};
-
-template <typename T, size_t N>
-struct TypeNameHelper<std::array<T, N> > {
-  static string GetTypeName() {
-    return "list of " + TypeTable::GetTypeName<T>();
-  }
-};
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/pipeline/data/types.h:390]: I can call it "list of ..." if you think it's better

**Comment by jantonguirao** [comment]
> Did the type name work at all for some runtime registered types?

It would work if you invoke the function

**Comment by klecki** [comment]
The extended type support probably deserves a test.

**Comment by JanuszL** [comment]
> The extended type support probably deserves a test.

I agree, a simple test for different data types would be great.

File: include/dali/core/convert.h [hunk 3]
```diff
@@ -167,23 +167,22 @@ DALI_HOST_DEV constexpr bool clamp(T value, ret_type<bool>) {
...
+  float f16_min = -65504.0f, f16_max = 65504.0f;
+  float f = clamp(value, ret_type<float>());
+  f = f < f16_min ? f16_min : f > f16_max ? f16_max : f;
+  return static_cast<float16>(f);
 }
 
 template <typename T>
 DALI_HOST_DEV constexpr T clamp(float16 value, ret_type<T>) {
-  return clamp(static_cast<float>(value), ret_type<T>());
+  return clamp(static_cast<float>(value), ret_type<T>{});
```
Comment (mzient) [inline include/dali/core/convert.h:178]: Nitpick: it doesn't change anything?

File: include/dali/core/convert.h [hunk 4]
```diff
@@ -167,23 +167,22 @@ DALI_HOST_DEV constexpr bool clamp(T value, ret_type<bool>) {
 
 template <typename T>
 DALI_HOST_DEV constexpr float16 clamp(T value, ret_type<float16>) {
-  return static_cast<float16>(value);
-}
-
-// __half does not have a constructor for int64_t, use long long
-DALI_HOST_DEV inline float16 clamp(int64_t value, ret_type<float16>) {
-  return static_cast<float16>(static_cast<long long int>(value));  // NOLINT
+  float f16_min = -65504.0f, f16_max = 65504.0f;
```
Comment (mzient) [inline include/dali/core/convert.h:170]: Nitpick
```suggestion
  constexpr float f16_min = -65504.0f, f16_max = 65504.0f;
```

File: include/dali/core/convert.h (hunk 3 above)
Comment (jantonguirao) [reply include/dali/core/convert.h:178]: true, it is just something I removed initially but I rewrote later

**Review by mzient — APPROVED** [review]

File: dali/test/python/test_operator_slice.py [hunk 5]
```diff
@@ -339,37 +350,40 @@ def check_slice_synth_data_vs_numpy(device, batch_size, input_shape, layout, axe
...
-                ((200,400), "HW", (1,), None),
-                ((200,400), "HW", None, "W"),
-                ((80, 30, 20, 3), "DHWC", (2,1,0), None),
-                ((80, 30, 20, 3), "DHWC", (0,1,2), None),
-                ((80, 30, 20, 3), "DHWC", (2,1), None),
-                ((80, 30, 20, 3), "DHWC", None, "WHD"),
-                ((80, 30, 20, 3), "DHWC", None, "DHW"),
-                ((80, 30, 20, 3), "DHWC", None, "WH"),
-                ((80, 30, 20, 3), "DHWC", None, "C")]:
+            for input_shape, layout, axes, axis_names, input_type, output_type in \
```
Comment (klecki) [inline dali/test/python/test_operator_slice.py:362]: If you want to not explode the number of cases, you can keep the old loop, select one of the cases and add after the loop another one:
```
for input_type in [types.FLOAT, types,UINT32, types.INT32, types.INT64]:
  for output_type in [None, types.FLOAT, types,UINT32, types.INT32, types.INT64]:
    yield check_slice_synth_data_vs_numpy device, batch_size, (200,400,3), "HWC", None, "HW", True, True, input_type, output_type
```
or something similar.

**Review by klecki — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**